cmd/squid-per-site-exporter/main.go

```diff
@@ -0,0 +1,298 @@
+			Name:    "squid_site_response_time_seconds",
+			Help:    "Response time per site in seconds",
+			Buckets: prometheus.DefBuckets,
+		},
+		[]string{"hostname"},
+	)
+)
+
+type SiteStats struct {
+	Hits     int64
+	Misses   int64
+	Bytes    int64
+	Requests int64
+}
+
+type Exporter struct {
+	mutex     sync.RWMutex
+	siteStats map[string]*SiteStats
+}
+
+func NewExporter() *Exporter {
+	return &Exporter{
+		siteStats: make(map[string]*SiteStats),
+	}
+}
+
+func (e *Exporter) parseLogLine(line string) {
+	// Squid log format: timestamp elapsedtime remotehost code/status bytes method URL rfc931 peerstatus/peerhost type
+	fields := strings.Fields(line)
+	if len(fields) < 7 {
+		return
+	}
+
+	// Extract relevant fields
+	elapsedTimeStr := fields[1]
+	codeStatus := fields[3]
+	bytesStr := fields[4]
+	method := fields[5]
+	urlStr := fields[6]
+
+	// Skip non-HTTP methods
+	if !strings.HasPrefix(method, "GET") && !strings.HasPrefix(method, "POST") &&
+		!strings.HasPrefix(method, "HEAD") && !strings.HasPrefix(method, "PUT") {
+		log.Printf("Unsupported method %q", method)
+		return
+	}
+
+	// Parse URL to extract hostname
+	parsedURL, err := url.Parse(urlStr)
+	if err != nil {
+		log.Printf("Invalid request URL %q: %v", urlStr, err)
+		return
+	}
+
+	hostname := parsedURL.Hostname()
+	if hostname == "" {
+		log.Printf("Missing hostname in URL %q", urlStr)
+		return
+	}
+
+	// Parse bytes
+	bytes, err := strconv.ParseInt(bytesStr, 10, 64)
+	if err != nil {
+		bytes = 0
+	}
+
+	// Parse elapsed time
+	elapsedTime, err := strconv.ParseFloat(elapsedTimeStr, 64)
+	if err != nil {
+		elapsedTime = 0
+	}
+
+	// Determine hit/miss from result code (token before '/')
+	// Consider only codes ending in "_HIT" as cache hits (e.g., TCP_HIT, MEM_HIT).
+	statusToken := codeStatus
+	if idx := strings.Index(codeStatus, "/"); idx >= 0 {
+		statusToken = codeStatus[:idx]
+	}
+	isHit := strings.HasSuffix(statusToken, "_HIT")
+
+	// Update stats
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+
+	if _, exists := e.siteStats[hostname]; !exists {
+		e.siteStats[hostname] = &SiteStats{}
+	}
+
+	stats := e.siteStats[hostname]
+	stats.Requests++
+	stats.Bytes += bytes
+
+	if isHit {
+		stats.Hits++
+	} else {
+		stats.Misses++
+	}
+
+	// Update Prometheus metrics
+	squidRequestsTotal.WithLabelValues(hostname).Inc()
+	squidBytesTotal.WithLabelValues(hostname).Add(float64(bytes))
+	squidResponseTime.WithLabelValues(hostname).Observe(elapsedTime / 1000.0) // Convert ms to seconds
+
+	if isHit {
+		squidHitTotal.WithLabelValues(hostname).Inc()
+	} else {
+		squidMissTotal.WithLabelValues(hostname).Inc()
+	}
+
+	// Ensure both hit and miss counters are initialized (even if 0) for this hostname
+	// This ensures squid_site_hits_total appears in metrics output even with 0 value
+	squidHitTotal.WithLabelValues(hostname).Add(0)
+	squidMissTotal.WithLabelValues(hostname).Add(0)
+
+	// Update hit ratio
+	if stats.Requests > 0 {
+		ratio := float64(stats.Hits) / float64(stats.Requests)
```

Comment:
I wonder whether it makes sense to use the values from the metric or from `stats`.
What happens if for some reason `stats.Hits` is out of sync with `squidHitTotal.WithLabelValues(hostname)`? The calculated ratio will not be in line with the values in the metrics.